src/datanode/src/sql/copy_table_from.rs

```diff
@@ -0,0 +1,254 @@
+
+                builder.endpoint(&s3.endpoint);
+
+                if s3.enable_virtual_host_style {
+                    builder.enable_virtual_host_style();
+                }
+                if let Some(region) = s3.region {
+                    builder.region(&region);
+                }
+                if let Some(access_key_id) = s3.access_key_id {
+                    builder.access_key_id(&access_key_id);
+                }
+                if let Some(key) = s3.secret_access_key {
+                    builder.secret_access_key(&key);
+                }
+                // TODO: Session_token
+
+                let accessor = builder.build().context(BuildObjectStoreSnafu)?;
+
+                let object_store = ObjectStore::new(accessor).finish();
+
+                (s3.uri.path, object_store)
+            }
+        };
+
+        Ok(CopyTableFromRequest {
+            catalog_name,
+            schema_name,
+            table_name,
+            source,
+            path,
+            object_store,
+        })
+    }
+
+    pub(crate) async fn copy_table_from(&self, req: CopyTableFromRequest) -> Result<Output> {
+        let table_ref = TableReference {
+            catalog: &req.catalog_name,
+            schema: &req.schema_name,
+            table: &req.table_name,
+        };
+        let table = self.get_table(&table_ref)?;
+
+        let reader = ParquetReader::new(req.object_store, &req.path, req.source);
+
+        let objects = reader.objects().await?;
+
+        let mut buf: Vec<RecordBatch> = Vec::new();
+
+        for obj in objects.iter() {
+            let reader = obj
+                .reader()
+                .await
+                .context(ReadObjectSnafu { path: &req.path })?;
+
+            let buf_reader = BufReader::new(reader.compat());
+
+            let builder = ParquetRecordBatchStreamBuilder::new(buf_reader)
+                .await
+                .context(ReadParquetSnafu)?;
+
+            ensure!(
+                *builder.schema() == table.schema().arrow_schema().clone(),
+                InvalidSchemaSnafu {
+                    table_schema: table.schema().arrow_schema().to_string(),
+                    file_schema: (*(builder.schema())).to_string()
+                }
+            );
+
+            let stream = builder
+                .build()
+                .context(BuildParquetRecordBatchStreamSnafu)?;
+
+            let chunk = stream
+                .try_collect::<Vec<_>>()
+                .await
+                .context(ReadParquetSnafu)?;
+
+            buf.extend(chunk.into_iter());
+        }
+
+        let fields = table
+            .schema()
+            .arrow_schema()
+            .fields()
+            .iter()
+            .map(|f| f.name().to_string())
+            .collect::<Vec<_>>();
+
+        // Vec<Columns>
+        let column_chunks = buf
+            .iter()
+            .map(|c| {
+                // TODO: how to remove the unwrap?
```

Comment:
We cannot simply unwrap here, because there are still some arrays in arrow that we don't support.